src/ast/irbuilderbpf.cpp

```diff
@@ -338,32 +338,85 @@ CallInst *IRBuilderBPF::CreateBpfPseudoCallValue(Map &map)
 
   assert(key->getType()->isPointerTy());
   FunctionType *lookup_func_type = FunctionType::get(
-      getInt8PtrTy(), { map_ptr->getType(), key->getType() }, false);
+      val_ptr_ty, { map_ptr->getType(), key->getType() }, false);
   PointerType *lookup_func_ptr_type = PointerType::get(lookup_func_type, 0);
   Constant *lookup_func = ConstantExpr::getCast(
       Instruction::IntToPtr,
       getInt64(libbpf::BPF_FUNC_map_lookup_elem),
       lookup_func_ptr_type);
-  return createCall(
-      lookup_func_type, lookup_func, { map_ptr, key }, "lookup_elem");
+  return createCall(lookup_func_type, lookup_func, { map_ptr, key }, name);
 }
 
-CallInst *IRBuilderBPF::CreateGetJoinMap(Value *ctx, const location &loc)
+CallInst *IRBuilderBPF::CreateGetJoinMap(BasicBlock *failure_callback,
+                                         const location &loc)
 {
-  AllocaInst *key = CreateAllocaBPF(getInt32Ty(), "key");
-  CreateStore(getInt32(0), key);
+  return createGetScratchMap(bpftrace_.maps[MapManager::Type::Join].value()->id,
+                             "join",
+                             getInt8PtrTy(),
+                             loc,
+                             failure_callback);
+}
+
+// createGetScratchMap will jump to failure_callback if it cannot find the map
+// value
+CallInst *IRBuilderBPF::createGetScratchMap(int mapid,
+                                            const std::string &name,
+                                            PointerType *val_ptr_ty,
+                                            const location &loc,
+                                            BasicBlock *failure_callback,
+                                            int key)
+{
+  AllocaInst *keyAlloc = CreateAllocaBPF(getInt32Ty(),
+                                         nullptr,
+                                         "lookup_" + name + "_key");
+  CreateStore(getInt32(key), keyAlloc);
 
   CallInst *call = createMapLookup(
-      bpftrace_.maps[MapManager::Type::Join].value()->id, key);
-  CreateHelperErrorCond(ctx, call, libbpf::BPF_FUNC_map_lookup_elem, loc, true);
+      mapid, keyAlloc, val_ptr_ty, "lookup_" + name + "_map");
+  CreateLifetimeEnd(keyAlloc);
+
+  Function *parent = GetInsertBlock()->getParent();
+  BasicBlock *lookup_success_block = BasicBlock::Create(
+      module_.getContext(), "lookup_" + name + "_success", parent);
+  BasicBlock *lookup_failure_block = BasicBlock::Create(
+      module_.getContext(), "lookup_" + name + "_failure", parent);
+  BasicBlock *lookup_merge_block = BasicBlock::Create(
+      module_.getContext(), "lookup_" + name + "_merge", parent);
+  Value *condition = CreateICmpNE(
+      CreateIntCast(call, getInt8PtrTy(), true),
+      ConstantExpr::getCast(Instruction::IntToPtr, getInt64(0), getInt8PtrTy()),
+      "lookup_" + name + "_cond");
+  CreateCondBr(condition, lookup_success_block, lookup_failure_block);
+
+  SetInsertPoint(lookup_success_block);
+  CreateBr(lookup_merge_block);
```

Comment:
Are there plans to add something to the `lookup_success_block`? Any reason why not to jump directly to `lookup_merge_block`? Or are we relying on LLVM to optimize this away?

Reply:
Nice, let me clean up this branch.